### PR TITLE
[DPE-5291] First Spark Test App version

### PR DESCRIPTION
**Pull request by deusebio**
This PR provides a first implementation for the Spark Test App, needed for the Data Warehousing with Spark demo. 

The `spark-test-app` allows to create an an simple app to be used in testing and demo. It now features two *flavour*

* vanilla: that simply runs a  count spark job to test functionality
* kafka: that connects to Kafka and perform some streaming aggregation 

The charm features the following integrations:

* Integration Hub (required) in order to create the spark service to be used to run the Spark Jobs
* Kafka (required only when running in Kafka flavour) to connect to Kafka
* PostgreSQL (required when persisting data into Hive Table)

Tests and CI are currently missing, as its development was done for the Spark Datawarehousing Demo, and they add later with a few know optimiation, among which:

* DeployMode cluster does not yet work in Kafka flavour
* KafkaApp should be refactored to use configuration files, instead of a long cmd line
* Allow autoinference of namespaces
* Allow deployment in constrained environments

### CODE REVIEW

**Comment by deusebio** [comment]
Commit [175776](https://github.com/canonical/spark-test-app/pull/2/commits/1757764f2dddfad04db12d04c9987bdfcfea25da) has been released on [Charmhub](https://charmhub.io/spark-test-app) as revision 1

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+### Deploy
+
+```bash
+# Create a model
+juju add-model dev
+# Enable DEBUG logging
+juju model-config logging-config="<root>=INFO;unit=DEBUG"
+# Deploy the charm
+juju deploy ./kafka-test-app_ubuntu-22.04-amd64.charm
```
Comment (theoctober19th) [inline CONTRIBUTING.md:57]: update `kafka-test-app` to `spark-test-app`.

File: icon.svg
Comment (theoctober19th) [inline icon.svg:1]: The logo is cool. Where do we take this from? :D

File: src/constants.py [hunk 2]
```diff
@@ -0,0 +1,14 @@
+#!/usr/bin/env python3
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Collection of globals common to the KafkaAppCharm."""
+
+CONTAINER="spark"
+KAFKA = "kafka"
```
Comment (theoctober19th) [inline src/constants.py:8]: suggestion: make the name more explainable (eg, `KAFKA_RELATION_NAME`)

File: src/core/domain.py [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+    extra_user_roles: str
+
+    @classmethod
+    @validator("extra_user_roles")
+    def _role_parser(cls, roles: str):
+        """Roles parsers."""
+        try:
+            _app_type = [AppType(value) for value in roles.split(",")]
+        except Exception as e:
+            raise ValidationError(f"could not properly parsed the roles configuration: {e}")
```
Comment (theoctober19th) [inline src/core/domain.py:141]:
```suggestion
            raise ValidationError(f"could not properly parse the roles configuration: {e}")
```

File: src/events/base.py [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+        # If the flavour Kafka is enabled, Kafka relation needs to exists
+        if (
+            self.context.config.flavour == Flavour.KAFKA.value and
+            not self.context.kafka
+        ):
+            return Status.MISSING_KAFKA.value
+
+        # If metastore is enabled, so need to be the object storage
+        if (
+            self.context.metastore and
```
Comment (theoctober19th) [inline src/events/base.py:52]: I did not understand how the existence of object storage relation is mandated due to metastore being enabled. Is this branch supposed to be equivalent to `self.context.config.flavour != Flavour.KAFKA.value`?

File: src/events/spark_app.py [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+            "EXTRA_ARGS": self.kafka_app.extra_args,
+            "SCRIPT": f"{self.workload.paths.bin_path}/{self.context.config.flavour.value}.py"
+        })
+
+    @compute_status
+    def _on_update_status(self, event: UpdateStatusEvent):
+        if kafka := self.context.kafka:
+            self.logger.info(f"Kafka data present")
+            self.logger.info(f"=> Username: {kafka.username}")
+            self.logger.info(f"=> Password: {kafka.password}")
```
Comment (theoctober19th) [inline src/events/spark_app.py:105]: We should probably not log this here.

File: src/workload.py [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+
+        self.container = container
+        self.user = user
+
+    @property
+    def _spark_layer(self):
+        """Return a dictionary representing a Pebble layer."""
+        layer = {
+            "summary": "spark history server layer",
+            "description": "pebble config layer for spark history server",
```
Comment (theoctober19th) [inline src/workload.py:43]: update description

File: CONTRIBUTING.md (hunk 1 above)
Comment (deusebio) [reply CONTRIBUTING.md:57]: huge catch! thanks!

File: icon.svg
Comment (deusebio) [reply icon.svg:1]: other repositories :D

File: src/events/base.py (hunk 4 above)
Comment (deusebio) [reply src/events/base.py:52]: when we enable the metastore, it means that we want to consolidate the results into a Hive Table and for storing the raw date we need to have access to an object storage backend (s3 or azure). 

I agree, it is somewhat obscure with the current comment, so I have explained this better in the comment to improve the clarity

File: src/events/spark_app.py (hunk 5 above)
Comment (deusebio) [reply src/events/spark_app.py:105]: yes, sorry, that was for testing :D Nice catch!

File: src/events/base.py (hunk 4 above)
Comment (theoctober19th) [reply src/events/base.py]: But even if we don't have PostgreSQL as metastore, we are still storing the raw data in the object storage backend, no? In that case, shouldn't the object storage be available at all times?

**Review by theoctober19th — APPROVED** [review]

File: actions.yaml [hunk 7]
```diff
@@ -0,0 +1,8 @@
+# Copyright 2023 Canonical Ltd.
```
Comment (welpaolo) [inline actions.yaml:1]: Copyright date

File: charmcraft.yaml [hunk 8]
```diff
@@ -0,0 +1,23 @@
+# Copyright 2023 Canonical Ltd.
```
Comment (welpaolo) [inline charmcraft.yaml:1]: Date

File: config.yaml [hunk 9]
```diff
@@ -0,0 +1,40 @@
+# Copyright 2023 Canonical Ltd.
```
Comment (welpaolo) [inline config.yaml:1]: Copyright date

File: metadata.yaml [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+
+containers:
+  spark:
+    resource: spark-image
+
+resources:
+  spark-image:
+    type: oci-image
+    description: OCI image for spark
+    upstream-source: ghcr.io/canonical/charmed-spark-kyuubi@sha256:1395fbcb34fee2ad1c939adc0e27d5b12fc10d67457f980df548612f735b860b
```
Comment (welpaolo) [inline metadata.yaml:39]: Can you put a comment with the spark/kyuubi version and revision?

File: pyproject.toml [hunk 11]
```diff
@@ -0,0 +1,101 @@
+# Copyright 2023 Canonical Ltd.
```
Comment (welpaolo) [inline pyproject.toml:1]: Copyright date

File: src/charm.py [hunk 12]
```diff
@@ -0,0 +1,45 @@
+#!/usr/bin/env python3
+# Copyright 2023 Canonical Ltd.
```
Comment (welpaolo) [inline src/charm.py:2]: here as well

File: src/common/relation/domain.py [hunk 13]
```diff
@@ -0,0 +1,38 @@
+"""Common domain objects."""
```
Comment (welpaolo) [inline src/common/relation/domain.py:1]:
```suggestion
#!/usr/bin/env python3

# Copyright 2024 Canonical Limited
# See LICENSE file for licensing details

"""Common domain objects."""
```

File: src/constants.py [hunk 14]
```diff
@@ -0,0 +1,14 @@
+#!/usr/bin/env python3
+# Copyright 2023 Canonical Ltd.
```
Comment (welpaolo) [inline src/constants.py:2]: Copyright date

File: src/core/domain.py [hunk 15]
```diff
@@ -0,0 +1,181 @@
...
+        return "SASL_PLAINTEXT" if self.tls is not None else "SASL_SSL"
+
+    @property
+    def bootstrap_server(self) -> str:
+        """Return the bootstrap servers endpoints."""
+        return self.endpoints
+
+
+class PostgreSQLProviderRelationDataBag(AuthDataBag):
+    """Class that handle the MongoDB relation databag."""
```
Comment (welpaolo) [inline src/core/domain.py:177]:
```suggestion
    """Class that handle the Postgresql relation databag."""
```

File: src/core/domain.py [hunk 16]
```diff
@@ -0,0 +1,181 @@
...
+
+    endpoints: str
+    username: str
+    password: str
+    tls: Optional[str] = None
+    tls_ca: Optional[str] = None
+
+
+class KafkaProviderRelationDataBag(AuthDataBag):
+    """Class for the provider relation databag."""
```
Comment (welpaolo) [inline src/core/domain.py:161]:
```suggestion
    """Class for the Kafka provider relation databag."""
```

File: src/core/workload.py [hunk 17]
```diff
@@ -0,0 +1,30 @@
+"""Module with the core interfaces for the Kafka app workload."""
```
Comment (welpaolo) [inline src/core/workload.py:1]:
```suggestion
#!/usr/bin/env python3

# Copyright 2024 Canonical Limited
# See LICENSE file for licensing details.

"""Module with the core interfaces for the Kafka app workload."""
```

File: src/events/__init__.py [hunk 18]
```diff
@@ -0,0 +1,22 @@
+"""Module for the base functionalities of event handlers."""
```
Comment (welpaolo) [inline src/events/__init__.py:1]:
```suggestion
#!/usr/bin/env python3

# Copyright 2024 Canonical Limited
# See LICENSE file for licensing details.

"""Module for the base functionalities of event handlers."""
```

File: src/events/spark_app.py [hunk 19]
```diff
@@ -0,0 +1,100 @@
...
+
+        src = source_path / source
+        dst = target_path / source
+
+        with src.open("r") as fid:
+            self.workload.write("\n".join(fid.readlines()), str(dst))
+
+    @compute_status
+    def _on_spark_pebble_ready(self, event: PebbleReadyEvent) -> None:
+        """Handle the bootstrap server changed."""
```
Comment (welpaolo) [inline src/events/spark_app.py:61]:
```suggestion
        """Handle the pebble ready event."""
```

File: src/events/spark_app.py [hunk 20]
```diff
@@ -0,0 +1,100 @@
...
+        # Event triggered when a bootstrap server was changed for this application
```
Comment (welpaolo) [inline src/events/spark_app.py:62]:
```suggestion
       
```

File: src/models.py [hunk 21]
```diff
@@ -0,0 +1,100 @@
+# Copyright 2023 Canonical Ltd.
```
Comment (welpaolo) [inline src/models.py:1]: Copyright date

File: tox.ini [hunk 22]
```diff
@@ -0,0 +1,76 @@
+# Copyright 2023 Canonical Ltd.
```
Comment (welpaolo) [inline tox.ini:1]: Copyright date

File: tox.ini [hunk 23]
```diff
@@ -0,0 +1,76 @@
...
+# See LICENSE file for licensing details.
+
+[tox]
+no_package = True
+skip_missing_interpreters = True
+env_list = lint, unit
+
+[vars]
+application = kafka-app
```
Comment (welpaolo) [inline tox.ini:10]:
```suggestion
application = spark-test-app
```

**Review by welpaolo — APPROVED** [review]
Looks good to me!
I left some minor comments

**Pull request merged**